File: go/rsa_lib.go
package main

import (
	"crypto/rand"
	"crypto/rsa"
	"crypto/x509"
	"encoding/pem"
	"fmt"
	"time"
)

// benchmarkKeyGeneration mede o tempo de geração de chaves usando biblioteca
func benchmarkKeyGeneration(bits int) time.Duration {
	start := time.Now()
	_, err := rsa.GenerateKey(rand.Reader, bits)
	if err != nil {
		panic(err)
	}
	return time.Since(start)
}

// benchmarkEncryption mede o tempo de criptografia usando biblioteca
func benchmarkEncryption(message []byte, publicKey *rsa.PublicKey) time.Duration {
	start := time.Now()
	_, err := rsa.EncryptPKCS1v15(rand.Reader, publicKey, message)
	if err != nil {
		panic(err)
	}
	return time.Since(start)
}

// benchmarkDecryption mede o tempo de descriptografia usando biblioteca
func benchmarkDecryption(ciphertext []byte, privateKey *rsa.PrivateKey) time.Duration {
	start := time.Now()
	_, err := rsa.DecryptPKCS1v15(rand.Reader, privateKey, ciphertext)
	if err != nil {
		panic(err)
	}
	return time.Since(start)
}

// main demonstra a implementação do RSA usando biblioteca
func main() {
	fmt.Println("=== Implementação RSA com Biblioteca ===")
	
	// Gerar chaves
	fmt.Println("Gerando chaves RSA...")
	keyGenerationTime := benchmarkKeyGeneration(2048)
	fmt.Printf("Tempo de geração de chaves: %v\n", keyGenerationTime)
	
	privateKey, err := rsa.GenerateKey(rand.Reader, 2048)
	if err != nil {
		panic(err)
	}
	
	publicKey := &privateKey.PublicKey
	fmt.Println("Chave gerada com sucesso!")
	
	// Mensagem de teste
	message := []byte("Hello, RSA!")
	fmt.Printf("Mensagem original: %s\n", string(message))
	
	// Criptografar
	encryptionTime := benchmarkEncryption(message, publicKey)
	ciphertext, err := rsa.EncryptPKCS1v15(rand.Reader, publicKey, message)
	if err != nil {
		panic(err)
	}
	fmt.Printf("Texto cifrado: %x\n", ciphertext)
	fmt.Printf("Tempo de criptografia: %v\n", encryptionTime)
	
	// Descriptografar
	decryptionTime := benchmarkDecryption(ciphertext, privateKey)
	decrypted, err := rsa.DecryptPKCS1v15(rand.Reader, privateKey, ciphertext)
	if err != nil {
		panic(err)
	}
	fmt.Printf("Texto descriptografado: %s\n", string(decrypted))
	fmt.Printf("Tempo de descriptografia: %v\n", decryptionTime)
	
	// Verificar se a descriptografia foi correta
	if string(message) == string(decrypted) {
		fmt.Println("✓ Criptografia/Descriptografia bem-sucedida!")
	} else {
		fmt.Println("✗ Erro na criptografia/Descriptografia!")
	}
	
	// Salvar chaves em formato PEM (opcional)
	fmt.Println("\n=== Informações das Chaves ===")
	
	// Chave pública
	publicKeyPEM := &pem.Block{
		Type:  "RSA PUBLIC KEY",
		Bytes: x509.MarshalPKCS1PublicKey(publicKey),
	}
	fmt.Println("Chave pública:")
	fmt.Println(string(pem.EncodeToMemory(publicKeyPEM)))
	
	// Chave privada
	privateKeyPEM := &pem.Block{
		Type:  "RSA PRIVATE KEY",
		Bytes: x509.MarshalPKCS1PrivateKey(privateKey),
	}
	fmt.Println("Chave privada:")
	fmt.Println(string(pem.EncodeToMemory(privateKeyPEM)))
} 

File: rust/src/rsa_lib.rs
use std::time::{Duration, Instant};
use rsa::{RsaPrivateKey, RsaPublicKey, pkcs8::{EncodePublicKey, LineEnding}};
use rsa::pkcs8::EncodePrivateKey;
use rsa::Pkcs1v15Encrypt;
use rand::rngs::OsRng;

// Estrutura para representar chaves RSA usando biblioteca
#[derive(Debug, Clone)]
pub struct RSALibKey {
    pub private_key: RsaPrivateKey,
    pub public_key: RsaPublicKey,
}

impl RSALibKey {
    pub fn new() -> Self {
        let private_key = RsaPrivateKey::new(&mut OsRng, 2048).expect("Falha ao gerar chave privada");
        let public_key = RsaPublicKey::from(&private_key);
        RSALibKey { private_key, public_key }
    }
}

// Função para criptografar usando biblioteca
pub fn encrypt_lib(message: &[u8], public_key: &RsaPublicKey) -> Result<Vec<u8>, rsa::Error> {
    public_key.encrypt(&mut OsRng, Pkcs1v15Encrypt, message)
}

// Função para descriptografar usando biblioteca
pub fn decrypt_lib(ciphertext: &[u8], private_key: &RsaPrivateKey) -> Result<Vec<u8>, rsa::Error> {
    private_key.decrypt(Pkcs1v15Encrypt, ciphertext)
}

// Função para medir o tempo de geração de chaves
pub fn benchmark_key_generation_lib() -> Duration {
    let start = Instant::now();
    let _key = RSALibKey::new();
    start.elapsed()
}

// Função para medir o tempo de criptografia
pub fn benchmark_encryption_lib(message: &[u8], key: &RSALibKey) -> Duration {
    let start = Instant::now();
    let _ciphertext = encrypt_lib(message, &key.public_key);
    start.elapsed()
}

// Função para medir o tempo de descriptografia
pub fn benchmark_decryption_lib(ciphertext: &[u8], key: &RSALibKey) -> Duration {
    let start = Instant::now();
    let _plaintext = decrypt_lib(ciphertext, &key.private_key);
    start.elapsed()
}

// Função principal para demonstração
pub fn main() {
    println!("=== Implementação RSA com Biblioteca ===");
    
    // Gerar chaves
    println!("Gerando chaves RSA...");
    let key_generation_time = benchmark_key_generation_lib();
    println!("Tempo de geração de chaves: {:?}", key_generation_time);
    
    let key = RSALibKey::new();
    println!("Chave gerada com sucesso!");
    
    // Mensagem de teste
    let message = b"Hello, RSA!";
    println!("Mensagem original: {}", String::from_utf8_lossy(message));
    
    // Criptografar
    let encryption_time = benchmark_encryption_lib(message, &key);
    let ciphertext = encrypt_lib(message, &key.public_key).expect("Falha na criptografia");
    println!("Texto cifrado: {:?}", ciphertext);
    println!("Tempo de criptografia: {:?}", encryption_time);
    
    // Descriptografar
    let decryption_time = benchmark_decryption_lib(&ciphertext, &key);
    let decrypted = decrypt_lib(&ciphertext, &key.private_key).expect("Falha na descriptografia");
    println!("Texto descriptografado: {}", String::from_utf8_lossy(&decrypted));
    println!("Tempo de descriptografia: {:?}", decryption_time);
    
    // Verificar se a descriptografia foi correta
    if message == decrypted.as_slice() {
        println!("✓ Criptografia/Descriptografia bem-sucedida!");
    } else {
        println!("✗ Erro na criptografia/Descriptografia!");
    }
    
    // Salvar chaves em formato PEM (opcional)
    println!("\n=== Informações das Chaves ===");
    println!("Chave pública:");
    println!("{}", key.public_key.to_public_key_pem(LineEnding::LF).unwrap());
    println!("Chave privada:");
    println!("{}", key.private_key.to_pkcs8_pem(LineEnding::LF).unwrap());
} 